src/kv/store.h

```diff
@@ -220,80 +220,112 @@ namespace kv
       return *result;
     }
 
-    void deserialize(const std::unique_ptr<AbstractSnapshot>& snapshot)
+    std::vector<uint8_t> serialise_snapshot(Version v) override
     {
-      std::lock_guard<SpinLock> mguard(maps_lock);
-
-      for (auto& map : maps)
-      {
-        map.second->lock();
-      }
-
-      const auto& snapshots = snapshot->get_map_snapshots();
-      CCF_ASSERT_FMT(
-        maps.size() == snapshots.size(),
-        "Number of maps does not match the snapshot, maps:{}, snapshots:{}",
-        maps.size(),
-        snapshots.size());
-      for (auto& s : snapshots)
-      {
-        auto search = maps.find(s->get_name());
-        if (search == maps.end())
-        {
-          throw ccf::ccf_logic_error(
-            fmt::format("Map does not exist - {}", s->get_name()));
-        }
-
-        search->second->apply(s);
-      }
-
-      for (auto& map : maps)
+      if (v < commit_version())
       {
-        map.second->unlock();
+        throw ccf::ccf_logic_error(fmt::format(
+          "Attempting to snapshot at version {} which is earlier than "
+          "committed version {} ",
+          v,
```

Comment:
```suggestion
          "Cannot snapshot at uncommitted version {}: earlier than "
          "committed version {} ",
```

Is more accurate in my opinion. "Attempting" isn't clearly an error, it may be confused for an in-progress operation, I think "Failed to", "Cannot" or something similarly definite would be preferable.

Reply:
Good point. I have re-phrased to `"Cannot snapshot at version {} which is earlier than committed version {}"` (slightly different from what you proposed as earlier version was probably committed).